abn/watcher/informer.go

```diff
@@ -19,29 +19,50 @@ type Iter8Watcher struct {
 }
 
 // NewIter8Watcher returns a watcher for iter8 related objects
-func NewIter8Watcher(resourceTypes []schema.GroupVersionResource, namespaces []string) *Iter8Watcher {
+// func NewIter8Watcher(resourceTypes []schema.GroupVersionResource, namespaces []string) *Iter8Watcher {
+func NewIter8Watcher(configFile string) *Iter8Watcher {
+	c := readServiceConfig(configFile)
+
 	w := &Iter8Watcher{
 		factories: map[string]dynamicinformer.DynamicSharedInformerFactory{},
 	}
 
-	handlerFunc := func(obj interface{}) {
-		wo := watchedObject{Obj: obj.(*unstructured.Unstructured)}
-		if precond(wo) {
-			handle(wo, resourceTypes, w.factories)
+	handlerFunc := func(action string, validNames []string, gvr schema.GroupVersionResource) func(obj interface{}) {
+		return func(obj interface{}) {
+			o := obj.(*unstructured.Unstructured)
+			if containsString(validNames, o.GetName()) {
+				handle(action, o, c, w.factories, gvr)
+			}
 		}
 	}
 
-	// for each namespace, resource type configure Informer
-	for _, ns := range namespaces {
+	// for each namespace:
+	//   identify all resources to watch (create informer for each), and
+	//   names of all resources to watch
+	//   create informerfactory
+	for ns, apps := range c {
+		var byResource map[schema.GroupVersionResource]([]string) = map[schema.GroupVersionResource]([]string){}
 		w.factories[ns] = dynamicinformer.NewFilteredDynamicSharedInformerFactory(k8sclient.Client.Dynamic(), 0, ns, nil)
-		for _, gvr := range resourceTypes {
+
+		// identify resources expected in namespace and list of resource names expected
+		// the resources will be used to create informers
+		// the list of names will be used to filter the objects that trigger the informer
+		for nm, details := range apps {
+			validNames := validObjectNames(nm, details.MaxNumCandidates)
+			for _, r := range details.Resources {
+				byResource[r.GroupVersionResource] = append(byResource[r.GroupVersionResource], validNames...)
+			}
+		}
+
+		// create informer for each resource in namespace
+		for gvr, validNames := range byResource {
 			informer := w.factories[ns].ForResource(gvr)
 			_, err := informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-				AddFunc: handlerFunc,
+				AddFunc: handlerFunc("ADD", validNames, gvr),
```

Comment:
Is it possible to use an enum here instead of a raw string?

Reply:
it isn't even needed